File: Leetcode_Medium/_235. Lowest Common Ancestor of a Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        # Start from the root node of the tree
        current = root
        
        # Traverse the tree
        while current:
            # If both p and q are greater than current node
            if p.val > current.val and q.val > current.val:
                # Go to the right subtree
                current = current.right
            # If both p and q are less than current node
            elif p.val < current.val and q.val < current.val:
                # Go to the left subtree
                current = current.left
            else:
                # We have found the split point, i.e. the LCA node
                return current
        

File: Leetcode_Medium/_235. Lowest Common Ancestor of a Binary Search Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

 public class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // Start from the root node of the tree
        TreeNode current = root;

        // Traverse the tree
        while (current != null) {
            // If both p and q are greater than current node
            if (p.val > current.val && q.val > current.val) {
                // Go to the right subtree
                current = current.right;
            } 
            // If both p and q are less than current node
            else if (p.val < current.val && q.val < current.val) {
                // Go to the left subtree
                current = current.left;
            } 
            // We have found the split point, i.e. the LCA node
            else {
                return current;
            }
        }
        return null; // In case the tree is empty
    }
}